### PR TITLE
Refactor controller logic

### PR DESCRIPTION
**Pull request by skonto**
- not sure if we should abandon glog,need to remove it if so from webhook etc
- I need to go through the logic for the quota feature plus the batch scheduler.
To test:
`make docker-build SPARK_IMAGE=lightbend/spark:2.1.1-OpenShift-2.4.3`
Then deploy as usual:

```
kubectl create namespace cert-manager
kubectl label namespace cert-manager certmanager.k8s.io/disable-validation=true
kubectl apply -f https://github.com/jetstack/cert-manager/releases/download/v0.8.1/cert-manager.yaml
# wait for the cert manager
kubectl apply -f all.yaml
# wait for the operator to come up
kubectl apply -f spark-app.yaml
```
Here is the log output of the working example I had:

```
+ [[ -z root:x:0:0:root:/root:/bin/ash ]]
+ exec /sbin/tini -s -- /usr/bin/spark-operator --metrics-addr=127.0.0.1:8080 --logtostderr --v=4 --enable-webhook=true
2019-09-16T21:04:58.202Z	INFO	main	Starting the Spark Operator
I0916 21:04:58.214226       9 crd.go:73] Creating CustomResourceDefinition sparkapplications.sparkoperator.k8s.io
I0916 21:04:58.734354       9 crd.go:73] Creating CustomResourceDefinition scheduledsparkapplications.sparkoperator.k8s.io
2019-09-16T21:04:59.244Z	INFO	main	Setting up the controller runtime manager
2019-09-16T21:04:59.597Z	INFO	controller-runtime.metrics	metrics server is starting to listen	{"addr": ":8080"}
2019-09-16T21:04:59.597Z	INFO	main	Registering Components.
2019-09-16T21:04:59.597Z	INFO	main	Adding Controllers.
2019-09-16T21:04:59.600Z	INFO	controller-runtime.controller	Starting EventSource	{"controller": "sparkapplication-controller", "source": "kind source: /, Kind="}
2019-09-16T21:04:59.600Z	INFO	controller-runtime.controller	Starting EventSource	{"controller": "sparkapplication-controller", "source": "kind source: /, Kind="}
2019-09-16T21:04:59.600Z	INFO	main	Setting up webhooks
2019-09-16T21:04:59.600Z	INFO	main	Getting the webhook server
2019-09-16T21:04:59.600Z	INFO	main	Starting the Cmd.
2019-09-16T21:04:59.601Z	INFO	controller-runtime.manager	starting metrics server	{"path": "/metrics"}
2019-09-16T21:04:59.701Z	INFO	controller-runtime.controller	Starting Controller	{"controller": "sparkapplication-controller"}
2019-09-16T21:04:59.706Z	INFO	controller-runtime.certwatcher	Updated current TLS certificate
2019-09-16T21:04:59.707Z	INFO	controller-runtime.certwatcher	Starting certificate watcher
2019-09-16T21:04:59.801Z	INFO	controller-runtime.controller	Starting workers	{"controller": "sparkapplication-controller", "worker count": 1}
2019-09-16T21:06:55.208Z	INFO	sparkapp-controller	Current Spark app state	{"appState": ""}
2019-09-16T21:06:55.208Z	INFO	sparkapp-controller	SparkApplication was added	{"namespace": "spark", "name": "spark-pi"}
2019-09-16T21:06:55.208Z	INFO	sparkapp-controller	spark-submit info	{"arguments": ["/opt/spark/bin/spark-submit", "--class", "org.apache.spark.examples.SparkPi", "--master", "k8s://https://10.96.0.1:443", "--deploy-mode", "cluster", "--conf", "spark.kubernetes.namespace=spark", "--conf", "spark.app.name=spark-pi", "--conf", "spark.kubernetes.driver.pod.name=spark-pi-driver", "--conf", "spark.kubernetes.container.image=lightbend/spark:2.1.2-OpenShift-2.4.4-rh-2.12", "--conf", "spark.kubernetes.container.image.pullPolicy=Always", "--conf", "spark.kubernetes.submission.waitAppCompletion=false", "--conf", "spark.kubernetes.driver.label.sparkoperator.k8s.io/app-name=spark-pi", "--conf", "spark.kubernetes.driver.label.sparkoperator.k8s.io/launched-by-spark-operator=true", "--conf", "spark.kubernetes.driver.label.sparkoperator.k8s.io/submission-id=649c3d61-315a-4b00-83e5-e9a015adc9d6", "--conf", "spark.driver.memory=500m", "--conf", "spark.kubernetes.authenticate.driver.serviceAccountName=spark-sa", "--conf", "spark.kubernetes.driver.label.version=2.4.3", "--conf", "spark.kubernetes.driver.secrets.mysecret=/etc/config-secret", "--conf", "spark.kubernetes.executor.label.sparkoperator.k8s.io/app-name=spark-pi", "--conf", "spark.kubernetes.executor.label.sparkoperator.k8s.io/launched-by-spark-operator=true", "--conf", "spark.kubernetes.executor.label.sparkoperator.k8s.io/submission-id=649c3d61-315a-4b00-83e5-e9a015adc9d6", "--conf", "spark.executor.instances=2", "--conf", "spark.executor.memory=500m", "--conf", "spark.kubernetes.executor.label.version=2.4.3", "local:///opt/spark/examples/jars/spark-examples_2.12-2.1.2-2.4.4-lightbend.jar", "100000000"]}
2019-09-16T21:06:55.208Z	DEBUG	controller-runtime.manager.events	Normal	{"object": {"kind":"SparkApplication","namespace":"spark","name":"spark-pi","uid":"51d1c98c-d6c2-4e14-b470-dfc058c21dfd","apiVersion":"sparkoperator.k8s.io/v1beta1","resourceVersion":"1467"}, "reason": "SparkApplicationAdded", "message": "SparkApplication spark-pi was added, enqueuing it for submission"}
2019-09-16T21:07:04.288Z	INFO	sparkapp-controller	spark-submit info	{"output": ""}

```

### CODE REVIEW

File: pkg/controller/sparkapplication/controller.go [hunk 1]
```diff
@@ -17,465 +17,506 @@ limitations under the License.
...
+	//		oldObject := e.ObjectOld.(*v1beta1.SparkApplication)
+	//		newObject := e.ObjectNew.(*v1beta1.SparkApplication)
+	//		if !reflect.DeepEqual (oldObject.Status, newObject.Status) {
+	//			// NO enqueue request
+	//			return false
+	//		}
+	//		// ENQUEUE request
+	//		return true
+	//	},
+	//}
```
Comment (skonto) [inline pkg/controller/sparkapplication/controller.go]: @yuchaoran2011 ideally we should progress based only on pod transition (real resources) and not status updates. Check here: https://github.com/kubernetes-sigs/kubebuilder/issues/618 on the other idempotency could be more desired. Haven't decided on this yet. Ideally I shouldnt get any changes that I dont want to, but I think I noticed some updated that is why I have this map `sm` check next to do a proper comparison.

File: pkg/controller/sparkapplication/controller.go [hunk 2]
```diff
@@ -17,465 +17,506 @@ limitations under the License.
...
-	c.enqueue(app)
+// ReconcileSparkApplication reconciles a SparkApplication object
+type ReconcileSparkApplication struct {
+	client           client.Client
+	scheme           *runtime.Scheme
+	metrics          *sparkAppMetrics
+	recorder         record.EventRecorder
+	crdClient        crdclientset.Interface
+	ingressURLFormat string
+	sm               sync.Map
```
Comment (skonto) [inline pkg/controller/sparkapplication/controller.go]: Keep previous app object here. Ideally we should get rid of this if we can transition without noise.

File: pkg/controller/sparkapplication/controller.go [hunk 3]
```diff
@@ -515,469 +556,324 @@ func shouldRetry(app *v1beta1.SparkApplication) bool {
...
-		c.recordSparkApplicationEvent(appToUpdate)
-		if err := c.validateSparkApplication(appToUpdate); err != nil {
-			appToUpdate.Status.AppState.State = v1beta1.FailedState
-			appToUpdate.Status.AppState.ErrorMessage = err.Error()
-		} else {
-			appToUpdate = c.submitSparkApplication(appToUpdate)
+	logger.Info("Current Spark app state", "appState", appToUpdate.Status.AppState.State)
+
+	// if spec has changed invalidate
+	oldApp, ok := r.sm.Load(request.NamespacedName)
```
Comment (skonto) [inline pkg/controller/sparkapplication/controller.go]: Check if spec has been invalidated else do the usual.

File: pkg/controller/sparkapplication/controller.go (hunk 1 above)
Comment (skonto) [reply pkg/controller/sparkapplication/controller.go]: Downside of keeping state is that on operator restarts we need to restore it. I havent done this yet.

Comment (skonto) [reply pkg/controller/sparkapplication/controller.go]: I will check if filtering out based on the request object can happen here.

File: Dockerfile [hunk 4]
```diff
@@ -16,22 +16,18 @@
 
 ARG SPARK_IMAGE=gcr.io/spark-operator/spark:v2.4.0
 
-FROM golang:1.12.5-alpine as builder
-ARG DEP_VERSION="0.5.3"
+FROM golang:1.12.9-alpine as builder
+ARG DEP_VERSION="0.5.4"
```
Comment (yuchaoran2011) [inline Dockerfile]: We no longer need `dep`

File: Dockerfile.rh [hunk 5]
```diff
@@ -25,18 +25,15 @@ ARG SPARK_IMAGE=gcr.io/spark-operator/spark:v2.4.0
 FROM golang:1.12.5-alpine as builder
 ARG DEP_VERSION="0.5.3"
```
Comment (yuchaoran2011) [inline Dockerfile.rh]: same here

**Review by yuchaoran2011 — APPROVED** [review]
LGTM. Let's not worry about resource quota for now.

**Pull request merged**